linz2osm/convert/osm.py

```diff
@@ -378,33 +432,72 @@ def export_custom(layer_in_dataset, feature_ids = None, workslice_id = None):
+            # Reuse a member feature if already included as a different role.
+            if key not in member_feature_map:
+                osm_feature_ids = _export_custom_data_row(writer, m_layer, member_tags[m.pk], m_processors, i, row_data, row_geom)
+                member_feature_map[key] = (osm_feature_type, osm_feature_ids)
+
     for i, (row_data, row_geom) in enumerate(data_table):
-        row_tags = []
-        for tag in layer_tags:
-            try:
-                v = tag.eval(row_data)
-            except tag.ScriptError, e:
-                emsg = "Error evaluating '%s' tag against record:\n" % tag
-                emsg += json.dumps(e.data, indent=2) + "\n"
-                emsg += str(e)
-                raise ValueError(emsg)
-            if (v is not None) and (v != ""):
-                row_tags.append((tag.tag, v, tag,))
-
-        # apply geometry processing
-        for p in processors:
-            row_geom = p.process(row_geom, fields=row_data, tags=row_tags, id=i)
-
-        if layer.special_node_reuse_logic:
-            first_node_ref = str(row_data.get(layer.special_start_node_field_name))
-            last_node_ref = str(row_data.get(layer.special_end_node_field_name))
-        else:
-            first_node_ref, last_node_ref = None, None
+        member_refs = []
+        for m in layer.members.all():
+            db_feature_id = row_data['member_feature_ids'][m.pk]
+            db_table_name = m.member_layer.name
+            key = (db_table_name, db_feature_id)
+            osm_feature_type, osm_feature_ids = member_feature_map[key]
+            for fid in osm_feature_ids:
+                member_refs.append((m.role, osm_feature_type, fid))
 
-        writer.add_feature(row_geom, row_tags, first_node_ref, last_node_ref)
+        _export_custom_data_row(writer, layer, layer_tags, processors, i, row_data, row_geom, member_refs=member_refs)
 
     return writer.xml()
 
+def _export_custom_data_row(writer, layer, tags, processors, i, row_data, row_geom, member_refs=None):
+    row_tags = []
+    for tag in tags:
+        try:
+            v = tag.eval(row_data)
+        except tag.ScriptError, e:
+            emsg = "Error evaluating '%s' tag against record:\n" % tag
+            emsg += json.dumps(e.data, indent=2) + "\n"
+            emsg += str(e)
+            raise ValueError(emsg)
+        if (v is not None) and (v != ""):
+            row_tags.append((tag.tag, v, tag,))
+
+    # apply geometry processing
+    for p in processors:
+        row_geom = p.process(row_geom, fields=row_data, tags=row_tags, id=i)
+
+    if layer.special_node_reuse_logic:
+        first_node_ref = str(row_data.get(layer.special_start_node_field_name))
+        last_node_ref = str(row_data.get(layer.special_end_node_field_name))
+    else:
+        first_node_ref, last_node_ref = None, None
+
+    if layer.geometry_type == "RELATION":
+        return writer.add_relation(row_tags, member_refs)
+    elif row_geom:
```

Comment:
No, it wasn't possible to have row_geom be None in the first place before: get_data_table would have errored out if there wasn't a wkb_geometry table.